src/sys/aio.rs

```diff
@@ -86,10 +118,21 @@ pub struct AioCb<'a> {
     /// Could this `AioCb` potentially have any in-kernel state?
     in_progress: bool,
     /// Used to keep buffers from Drop'ing
-    keeper: Keeper<'a>
+    buffer: Buffer<'a>
 }
 
 impl<'a> AioCb<'a> {
+    /// Remove the inner `Buffer` and return it
+    ///
+    /// It is an error to call this method while the `AioCb` is still in
+    /// progress.
+    pub fn buffer(&mut self) -> Buffer<'a> {
```

Comment:
Does it make sense to make this return a `Result` type instead of having failing if the in progress invariant is broken?

Reply:
I think not.  If we do that, then `AioCb::into_buffer` will need to be able to handle it, such as by returning `None`.  However, that method can already return `Buffer::none`, which means something different.  It would be confusing to return two different None types.

In my experience, your program's flow should make it obvious whether any given `AioCb` is still in progress or not.  Treating an in-progress `AioCb` as though it has already finished is almost always going to be a programming bug.  So I think `assert`ing is appropriate.